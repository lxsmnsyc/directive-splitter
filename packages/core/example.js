import { compile } from 'dismantle';

const code = `
import { server$ } from 'my-example';

const foo = 'foo';
const bar = 'bar';

function log(prefix, suffix) {
  console.log(prefix, suffix);
}

const logPrefix = server$((prefix) => {
  if (prefix === 'foo') {
    log(prefix, foo);
  } else {
    log(prefix, bar);
  }
});
`;

const result = await compile(code, '/path/to/example.ts', {
  key: 'example',
  mode: 'server',
  env: 'development',
  directives: [
    {
      value: 'use server',
      import: {
        kind: 'named',
        name: '$$server',
        source: 'my-example',
      },
    },
  ],
  functions: [
    {
      source: {
        kind: 'named',
        name: 'server$',
        source: 'my-example',
      },
      target: {
        kind: 'named',
        name: 'server$',
        source: 'my-example/server',
      },
    },
  ],
});

console.log('ENTRY');
console.log(result.code);
console.log(result.map);

console.log('\nVIRTUALS');
for (const [name, content] of result.files) {
  console.log('FILE', name);
  console.log(content);
}

console.log('\nENTRIES');
console.log(result.entries);

console.log('\nROOTs');
console.log(result.roots);
