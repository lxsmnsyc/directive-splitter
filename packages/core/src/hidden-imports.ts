import type { ImportDefinition } from './types';

export const HIDDEN_FUNC: ImportDefinition = {
  kind: 'named',
  source: 'dismantle/runtime',
  name: '$$func',
};

export const HIDDEN_GENERATOR: ImportDefinition = {
  kind: 'named',
  source: 'dismantle/runtime',
  name: '$$gen',
};
