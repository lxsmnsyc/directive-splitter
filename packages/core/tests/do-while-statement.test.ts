import { describe, it, expect } from 'vitest';
import * as compiler from '../src';
import { ID, CLIENT, SERVER } from './example';

describe('DoWhileStatement', () => {
  describe('client', () => {
    it('should transform valid server do-while statements', async () => {
      const code = `
      do {
        'use server';
        await doStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should skip server do-while statements in non-async functions', async () => {
      const code = `
      const example = () => {
        do {
          'use server';
          doStuff();
        } while (cond())
      };
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should transform valid server functions with scope', async () => {
      const code = `
      async function foo() {
        const value = 'foo bar';
        do {
          'use server';
          await doStuff(value);
        } while (cond())
      }
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should skip top-level values for scope', async () => {
      const code = `
      const value = 'foo bar';
      do {
        'use server';
        await doStuff(value);
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should transform break statements', async () => {
      const code = `
      do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          break;
        }
        await doMoreStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should transform continue statements', async () => {
      const code = `
      do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          continue;
        }
        await doMoreStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should transform labeled break statements', async () => {
      const code = `
      foo: do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          break foo;
        }
        await doMoreStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should transform labeled continue statements', async () => {
      const code = `
      foo: do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          continue foo;
        }
        await doMoreStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
  });
  describe('server', () => {
    it('should transform valid server do-while statements', async () => {
      const code = `
      do {
        'use server';
        await doStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should skip server do-while statements in non-async functions', async () => {
      const code = `
      const example = () => {
        do {
          'use server';
          doStuff();
        } while (cond())
      };
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should transform valid server functions with scope', async () => {
      const code = `
      async function foo() {
        const value = 'foo bar';
        do {
          'use server';
          await doStuff(value);
        } while (cond())
      }
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should skip top-level values for scope', async () => {
      const code = `
      const value = 'foo bar';
      do {
        'use server';
        await doStuff(value);
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should transform break statements', async () => {
      const code = `
      do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          break;
        }
        await doMoreStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should transform continue statements', async () => {
      const code = `
      do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          continue;
        }
        await doMoreStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should transform labeled break statements', async () => {
      const code = `
      foo: do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          break foo;
        }
        await doMoreStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should transform labeled continue statements', async () => {
      const code = `
      foo: do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          continue foo;
        }
        await doMoreStuff();
      } while (cond())
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
  });
});
