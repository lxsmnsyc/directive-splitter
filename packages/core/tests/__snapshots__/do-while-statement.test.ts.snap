// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DoWhileStatement > client > should skip server do-while statements in non-async functions 1`] = `
{
  "code": "const example = () => {
  do {
    'use server';
    doStuff();
  } while (cond());
};",
  "entries": [],
  "files": Map {},
  "map": {
    "file": undefined,
    "mappings": "AACM,MAAMA,OAAO,GAAGA,CAAA,KAAM;EACpB,GAAG;IACD,YAAY;IACZC,OAAO,CAAC,CAAC;EACX,CAAC,QAAQC,IAAI,CAAC,CAAC;AACjB,CAAC",
    "names": [
      "example",
      "doStuff",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const example = () => {
        do {
          'use server';
          doStuff();
        } while (cond())
      };
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`DoWhileStatement > client > should skip top-level values for scope 1`] = `
{
  "code": "import { value } from "./example.ts?directive=0.ts";
do {
  const _block = (await import("./example.ts?directive=2.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=2.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export const value = 'foo bar';",
    "/path/to/example.ts?directive=1.ts" => "export default (async function () {
  try {
    await doStuff(value);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=2.ts" => "import { server as _entry } from "directive-splitter/example-client";
export default _entry("26ecc3a5-0-anonymous");",
  },
  "map": {
    "file": undefined,
    "mappings": ";AAEM;EAAA,MAAAA,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAGSE,IAAI,CAAC,CAAC",
    "names": [
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const value = 'foo bar';
      do {
        'use server';
        await doStuff(value);
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=1.ts",
  ],
}
`;

exports[`DoWhileStatement > client > should transform break statements 1`] = `
{
  "code": "do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 0) {
    break;
  } else if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    if (cond()) {
      await doStuff(value);
    } else {
      return [0, null];
    }
    await doMoreStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-client";
export default _entry("26ecc3a5-0-anonymous");",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM;EAAA,MAAAA,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA;EAAA,WAAAA,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAQSE,IAAI,CAAC,CAAC",
    "names": [
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          break;
        }
        await doMoreStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > client > should transform continue statements 1`] = `
{
  "code": "do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 1) {
    continue;
  } else if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    if (cond()) {
      await doStuff(value);
    } else {
      return [1, null];
    }
    await doMoreStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-client";
export default _entry("26ecc3a5-0-anonymous");",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM;EAAA,MAAAA,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA;EAAA,WAAAA,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAQSE,IAAI,CAAC,CAAC",
    "names": [
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          continue;
        }
        await doMoreStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > client > should transform labeled break statements 1`] = `
{
  "code": "foo: do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 0) {
    if (_result === "foo") {
      break foo;
    }
  } else if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    if (cond()) {
      await doStuff(value);
    } else {
      return [0, "foo"];
    }
    await doMoreStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-client";
export default _entry("26ecc3a5-0-anonymous");",
  },
  "map": {
    "file": undefined,
    "mappings": "AACMA,GAAG,EAAE;EAAA,MAAAC,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,IAAAC,OAAA;MAAA,MAAAJ,GAAA;IAAA;EAAA,WAAAG,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAQIE,IAAI,CAAC,CAAC",
    "names": [
      "foo",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      foo: do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          break foo;
        }
        await doMoreStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > client > should transform labeled continue statements 1`] = `
{
  "code": "foo: do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 1) {
    if (_result === "foo") {
      continue foo;
    }
  } else if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    if (cond()) {
      await doStuff(value);
    } else {
      return [1, "foo"];
    }
    await doMoreStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-client";
export default _entry("26ecc3a5-0-anonymous");",
  },
  "map": {
    "file": undefined,
    "mappings": "AACMA,GAAG,EAAE;EAAA,MAAAC,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,IAAAC,OAAA;MAAA,SAAAJ,GAAA;IAAA;EAAA,WAAAG,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAQIE,IAAI,CAAC,CAAC",
    "names": [
      "foo",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      foo: do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          continue foo;
        }
        await doMoreStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > client > should transform valid server do-while statements 1`] = `
{
  "code": "do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    await doStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-client";
export default _entry("26ecc3a5-0-anonymous");",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM;EAAA,MAAAA,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAGSE,IAAI,CAAC,CAAC",
    "names": [
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      do {
        'use server';
        await doStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > client > should transform valid server functions with scope 1`] = `
{
  "code": "async function foo() {
  const value = 'foo bar';
  do {
    const _block = (await import("./example.ts?directive=1.ts")).default;
    const [_type, _result, _mutations] = await _block(value);
    if (_type === 4) {
      throw _result;
    }
  } while (cond());
}",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function (value) {
  try {
    await doStuff(value);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-client";
export default _entry("26ecc3a5-0-foo");",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM,eAAeA,GAAGA,CAAA,EAAG;EACnB,MAAMC,KAAK,GAAG,SAAS;EACvB;IAAA,MAAAC,MAAA,iDAAAC,OAAA;IAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA,CADMD,KAAK;IAAA,IAAAG,KAAA;MAAA,MAAAC,OAAA;IAAA;EAAA,SAIFE,IAAI,CAAC,CAAC;AACjB",
    "names": [
      "foo",
      "value",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      async function foo() {
        const value = 'foo bar';
        do {
          'use server';
          await doStuff(value);
        } while (cond())
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > server > should skip server do-while statements in non-async functions 1`] = `
{
  "code": "const example = () => {
  do {
    'use server';
    doStuff();
  } while (cond());
};",
  "entries": [],
  "files": Map {},
  "map": {
    "file": undefined,
    "mappings": "AACM,MAAMA,OAAO,GAAGA,CAAA,KAAM;EACpB,GAAG;IACD,YAAY;IACZC,OAAO,CAAC,CAAC;EACX,CAAC,QAAQC,IAAI,CAAC,CAAC;AACjB,CAAC",
    "names": [
      "example",
      "doStuff",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const example = () => {
        do {
          'use server';
          doStuff();
        } while (cond())
      };
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`DoWhileStatement > server > should skip top-level values for scope 1`] = `
{
  "code": "import { value } from "./example.ts?directive=0.ts";
do {
  const _block = (await import("./example.ts?directive=2.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=2.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export const value = 'foo bar';",
    "/path/to/example.ts?directive=1.ts" => "import { value } from "./example.ts?directive=0.ts";
export default (async function () {
  try {
    await doStuff(value);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=2.ts" => "import { server as _entry } from "directive-splitter/example-server";
import _root from "./example.ts?directive=1.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": ";AAEM;EAAA,MAAAA,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAGSE,IAAI,CAAC,CAAC",
    "names": [
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const value = 'foo bar';
      do {
        'use server';
        await doStuff(value);
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=1.ts",
  ],
}
`;

exports[`DoWhileStatement > server > should transform break statements 1`] = `
{
  "code": "do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 0) {
    break;
  } else if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    if (cond()) {
      await doStuff(value);
    } else {
      return [0, null];
    }
    await doMoreStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-server";
import _root from "./example.ts?directive=0.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM;EAAA,MAAAA,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA;EAAA,WAAAA,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAQSE,IAAI,CAAC,CAAC",
    "names": [
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          break;
        }
        await doMoreStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > server > should transform continue statements 1`] = `
{
  "code": "do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 1) {
    continue;
  } else if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    if (cond()) {
      await doStuff(value);
    } else {
      return [1, null];
    }
    await doMoreStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-server";
import _root from "./example.ts?directive=0.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM;EAAA,MAAAA,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA;EAAA,WAAAA,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAQSE,IAAI,CAAC,CAAC",
    "names": [
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          continue;
        }
        await doMoreStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > server > should transform labeled break statements 1`] = `
{
  "code": "foo: do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 0) {
    if (_result === "foo") {
      break foo;
    }
  } else if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    if (cond()) {
      await doStuff(value);
    } else {
      return [0, "foo"];
    }
    await doMoreStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-server";
import _root from "./example.ts?directive=0.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": "AACMA,GAAG,EAAE;EAAA,MAAAC,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,IAAAC,OAAA;MAAA,MAAAJ,GAAA;IAAA;EAAA,WAAAG,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAQIE,IAAI,CAAC,CAAC",
    "names": [
      "foo",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      foo: do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          break foo;
        }
        await doMoreStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > server > should transform labeled continue statements 1`] = `
{
  "code": "foo: do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 1) {
    if (_result === "foo") {
      continue foo;
    }
  } else if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    if (cond()) {
      await doStuff(value);
    } else {
      return [1, "foo"];
    }
    await doMoreStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-server";
import _root from "./example.ts?directive=0.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": "AACMA,GAAG,EAAE;EAAA,MAAAC,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,IAAAC,OAAA;MAAA,SAAAJ,GAAA;IAAA;EAAA,WAAAG,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAQIE,IAAI,CAAC,CAAC",
    "names": [
      "foo",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      foo: do {
        'use server';
        if (cond()) {
          await doStuff(value);
        } else {
          continue foo;
        }
        await doMoreStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > server > should transform valid server do-while statements 1`] = `
{
  "code": "do {
  const _block = (await import("./example.ts?directive=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 4) {
    throw _result;
  }
} while (cond());",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function () {
  try {
    await doStuff();
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-server";
import _root from "./example.ts?directive=0.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM;EAAA,MAAAA,MAAA,iDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA,SAGSE,IAAI,CAAC,CAAC",
    "names": [
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      do {
        'use server';
        await doStuff();
      } while (cond())
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;

exports[`DoWhileStatement > server > should transform valid server functions with scope 1`] = `
{
  "code": "async function foo() {
  const value = 'foo bar';
  do {
    const _block = (await import("./example.ts?directive=1.ts")).default;
    const [_type, _result, _mutations] = await _block(value);
    if (_type === 4) {
      throw _result;
    }
  } while (cond());
}",
  "entries": [
    "/path/to/example.ts?directive=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?directive=0.ts" => "export default (async function (value) {
  try {
    await doStuff(value);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?directive=1.ts" => "import { server as _entry } from "directive-splitter/example-server";
import _root from "./example.ts?directive=0.ts";
export default _entry("26ecc3a5-0-foo", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM,eAAeA,GAAGA,CAAA,EAAG;EACnB,MAAMC,KAAK,GAAG,SAAS;EACvB;IAAA,MAAAC,MAAA,iDAAAC,OAAA;IAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA,CADMD,KAAK;IAAA,IAAAG,KAAA;MAAA,MAAAC,OAAA;IAAA;EAAA,SAIFE,IAAI,CAAC,CAAC;AACjB",
    "names": [
      "foo",
      "value",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
      "cond",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      async function foo() {
        const value = 'foo bar';
        do {
          'use server';
          await doStuff(value);
        } while (cond())
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?directive=0.ts",
  ],
}
`;
